src/controllers/authController.js

```diff
@@ -13,31 +14,47 @@ const register = async (req, res) => {
     throw new CustomError.BadRequestError('Please provide all required fields');
   }
 
-  const emailAlreadyExists = await User.findOne({ email }); // Check if a user with the email already exists
-  if (emailAlreadyExists) {
-    throw new CustomError.BadRequestError('Email already exists'); // If user exists, throw an error
-  }
-
-  const isFirstAccount = (await User.countDocuments({})) === 0; // Check if it's the first account
-  const role = isFirstAccount ? 'admin' : 'user'; // Assign a role based on first account or not
-
-  const user = await User.create({
-    name,
-    username,
-    email,
-    password,
-    role,
-  }); // Create a new user in the database
+  const session = await mongoose.startSession();
+  session.startTransaction();
 
-  // Send the welcome email
-  await emailSender.sendWelcomeEmail(user.email, user);
+  try {
+    const emailAlreadyExists = await User.findOne({ email }); // Check if a user with the email already exists
 
-  const tokenUser = createTokenUser(user); // Create a token based on user data
-  attachCookiesToResponse({ res, user: tokenUser }); // Attach the token to cookies and send in the response
-
-  res.status(StatusCodes.CREATED).json({ user: tokenUser }); // Send a successful response with user data
+    if (emailAlreadyExists) {
+      throw new CustomError.BadRequestError('Email already exists'); // If user exists, throw an error
+    }
 
-  /* 
+    const isFirstAccount = (await User.countDocuments({})) === 0; // Check if it's the first account
+    const role = isFirstAccount ? 'admin' : 'user'; // Assign a role based on first account or not
+
+    const user = await User.create(
+      [
+        {
+          name,
+          username,
+          email,
+          password,
+          role,
+        },
+      ],
+      { session }
+    ); // Create a new user in the database
+
+    // Send the welcome email
+    await emailSender.sendWelcomeEmail(user[0].email, user[0]);
+
+    const tokenUser = createTokenUser(user[0]); // Create a token based on user data
```

Comment:
Did you run into some issue that led you to change  the input of the `User.create` function from an object to an array? It seems like it makes things a bit more unclear because now `user` is a _list_ containing one user and we have to access it the user that we created by index `[0]` everywhere below.

Reply:
Oh never mind, I see now this message when I try it without the array:

```
(node:50922) [MONGOOSE] Warning: WARNING: to pass a `session` to `Model.create()` in Mongoose, you **must** pass an array as the first argument. See: https://mongoosejs.com/docs/api/model.html#Model.create()
```

TIL (today I learned) !

Reply:
In that case, I might suggest using array destructuring to assign the first value in the list returned (which is the user that we've just created) to the `user` variable like this: [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)

`const [user] = await User.create([{...}])`

Reply:
distructuring done commit b9f612dfa6ee96f11f60b9556720699d2eebd226

Reply:
To clarify, this destructuring is only needed where we do `User.create([ {...}])` because we are passing in an _array_ to the `User.create` function which means that mongoose will return an _array_ of results. But we only need the first user in the array, so we destrucutre. 

But in the other place that you changed in that commit in the `login` function, this is not needed because `findOne` always returns just a single object and never an array.

Reply:
Ah that's fine I probably just mixed up the two places when writing the comment; everything looks good now :+1: